tests/Unit/DataStructures/Tensor/Expressions/Test_Divide.cpp

```diff
@@ -0,0 +1,253 @@
+  // Use explicit type (vs auto) for LHS Tensor so the compiler checks the
+  // return type of `evaluate`
+  const tnsr::ii<DataType, dim> Lij_from_Sij_over_R =
+      TensorExpressions::evaluate<ti_i, ti_j>(S(ti_i, ti_j) / 4.3);
+  // \f$L_{ij} = (R * S_{ij} / T / G)\f$
+  const tnsr::ii<DataType, dim> Lij_from_R_Sij_over_T =
+      TensorExpressions::evaluate<ti_i, ti_j>((-5.2 * S(ti_i, ti_j) / 1.6) /
+                                              2.1);
+
+  for (size_t i = 0; i < dim; i++) {
+    for (size_t j = 0; j < dim; j++) {
+      CHECK_ITERABLE_APPROX(Lij_from_Sij_over_R.get(i, j), S.get(i, j) / 4.3);
+      CHECK_ITERABLE_APPROX(Lij_from_R_Sij_over_T.get(i, j),
+                            -5.2 * S.get(i, j) / 1.6 / 2.1);
+    }
+  }
+}
+
+// \brief Test the division of a tensor expression over a rank 0 tensor
+// expression is correctly evaluated
+//
+// \details
+// The cases tested are:
+// - \f$L_{i}{}^{j} = S^{i}{}_{j} / R\f$
+// - \f$L^{k}{}_{i} = (R T_{i}{}^{k}) / (T_{j}{}^{l} S^{j}{}_{l})\f$
+// - \f$L_{i}{}^{k} = T_{i}{}^{k} / R^2 / R\f$
+//
+// where \f$R\f$, \f$S\f$, \f$T\f$, and \f$L\f$ are Tensors with data type
+// `double` or DataVector.
+//
+// \tparam DataType the type of data being stored in the operands and result
+template <typename DataType, typename Generator>
+void test_divide_rank0(const DataType& used_for_size,
+                       const gsl::not_null<Generator*> generator) noexcept {
+  std::uniform_real_distribution<> distribution(-2.0, 2.0);
+  constexpr size_t dim = 3;
+
+  const auto R = make_with_random_values<Scalar<DataType>>(
+      generator, distribution, used_for_size);
+  const auto S = make_with_random_values<tnsr::Ij<DataType, dim>>(
+      generator, distribution, used_for_size);
+  const auto T = make_with_random_values<tnsr::iJ<DataType, dim>>(
+      generator, distribution, used_for_size);
+
+  // \f$L_{i}{}^{j} = S^{i}{}_{j} / R\f$
+  // Use explicit type (vs auto) for LHS Tensor so the compiler checks the
+  // return type of `evaluate`
+  const tnsr::iJ<DataType, dim> result1 =
+      TensorExpressions::evaluate<ti_j, ti_I>(S(ti_I, ti_j) / R());
+
+  for (size_t i = 0; i < dim; i++) {
+    for (size_t j = 0; j < dim; j++) {
+      CHECK(result1.get(j, i) == S.get(i, j) / get(R));
+    }
+  }
+
+  // \f$L^{k}{}_{i} = (R T_{i}{}^{k}) / (T_{j}{}^{l} S^{j}{}_{l})\f$
+  const tnsr::Ij<DataType, dim> result2 =
+      TensorExpressions::evaluate<ti_K, ti_i>(
+          (R() * T(ti_i, ti_K)) / ((T(ti_j, ti_L) * S(ti_J, ti_l))));
+
+  DataType result2_expected_denominator =
+      make_with_value<DataType>(used_for_size, 0.0);
+  for (size_t j = 0; j < dim; j++) {
+    for (size_t l = 0; l < dim; l++) {
+      result2_expected_denominator += T.get(j, l) * S.get(j, l);
+    }
+  }
+
+  for (size_t i = 0; i < dim; i++) {
+    for (size_t k = 0; k < dim; k++) {
+      CHECK_ITERABLE_APPROX(
+          result2.get(k, i),
+          get(R) * T.get(i, k) / result2_expected_denominator);
```

Comment:
this test is failing, though it's not obvious to me why. The math is slightly different in terms of parentheses, but I'm not sure that's the cause since that should only be rounding errors. I'm also not sure we want to allow this type of math. Doesn't this mean `T_{j}{}^{l} S^{j}{}_{l}` is computed for _every_ LHS index instead of just once? It's probably better to require the `/` rhs be `Scalar<DataType>` (or the expression corresponding to a `Scalar<DataType>`). Considering you as the author of the TensorExpressions might not have noticed this huge performance issue, it's probably best to prevent it completely. Given that change, I'm not reviewing code below in anticipation some of it will change :)

Reply:
This is a really good point, and I think it's good to also point out that this is an issue that is not unique to this in-progress `/` operation. For example, all of the following would also suffer from unnecessary repeated calculation:
- `R(ti_a) * (S(ti_B) * T(ti_b))`
- `R(ti_a) - G(ti_a) * sqrt(S(ti_B) * T(ti_b))`
- `R(ti_a) * (S(ti_t, ti_t) + T(ti_t))`

I agree with wanting to avoid the performance issue you've pointed out. However, where do we draw the line with what we permit for the denominator? For example, these denominators are all rank 0 TE types that are not rank 0 `TensorAsExpression` (the TE type for single `Tensor`s):
- ` / (R() * S())` : rank 0 `OuterProduct`, Marlo needs to do this operation.
- `/ (R(ti_t, ti_t) + S(ti_T))` : rank 0 `AddSub`
- `/ sqrt(R())` : rank 0 `SquareRoot`
- `/ sqrt(R(ti_t) - S(ti_t))` : rank 0 `SquareRoot`

So I guess:
- This performance problem wouldn't be unique to `/` with rank 0 TEs, so maybe this is an issue to look into solving across the board.
- What is the extent of the types of denominators we would accept, given some of the examples above? I'm not sure where to draw the line to balance ease of use with what the human should be expected to do up front. Something else to maybe consider:
    -  For a case like `T(ti_c) / (R(ti_A, ti_B) * S(ti_a, ti_b))`, I'm guessing that for the for loop way, an author would be asked to write a separate series of loops to compute this denominator before the separate loop for `c`. And if that's the case, then it seems equally reasonable to ask someone to write something like `evaluate(denominator, 1.0 / (R(ti_A, ti_B) * S(ti_a, ti_b)))`, and then use `denominator` in `evaluate<ti_c>(result, T(ti_c) * denominator())`

Let me know your thoughts!